First pull request: This is a  proposal for the structure of the code.

R-Oc-A opened the pull request:
This is an incomplete version of the code.
Several changes may be included in the following weeks. 

The status of the project is the following: I've finished the pulstar version of the code, as in I have a running code that projects the pulsation, rotation, temperature and Log g on a raster sphere as well as the visibility of each cell. I've still left to write the functions and/or methods to write the data of the grid that shall be used by the profile code to produce the light curves. 

The following week I'll be exploring  the "polars" crate to read and write data for this project. 

The main thing I'd like for you to check is that the information printed by the computation report of the pulstar code is Physically correct and if it's not the case, to know where you might think I should look. 
To do this you can run the binary on the pulstar directory with the following command:

`cargo run -- input_pulstar.txt input_timepoints.txt`

Also, I want to ask for suggestions on naming of libraries and modules. The main library is called as of now as "temp_name_lib" and the some mathematical functions and methods are grouped in a module called "joris_math". This and the overall structure of the code can be discussed once you return to Leuven.

R-Oc-A committed:
Testing a workspace modular structure for the project

R-Oc-A committed:
**Follow up on the structure. As of now: two binary packages and one library package within a workspace.**
The contents of the code will be divided in three packages contained on a workspace.
The packages are the following:
The pulstar binary
The profile binary
The temp_lib_name library

On this first iteration of the project, I shall work on the temp_lib_name library.
This library is divided on two modules. joris_math module and star_physics module.
It's important to tell that I have yet to define the following:

1) The access of modules, structures and functions.
   Right now I'm declaring everything as public. But I'll figure that out later.
2) Tests and documentation.
   There's only a test for the associated Legendre polynomials
   which follows the structure of assert_eq!(plmn(args),expected_val).
   Might be wise to ask Joris about this. As for the documentation,
   it seems wise to follow the same structure as Joris.

R-Oc-A committed:
Will be working on changing some functions to use references as they do not need to copy them.

R-Oc-A committed:
Intermediate commit so that I don't loose progress.

R-Oc-A committed:
Continuing work on the structure. Finished writing the displacement module

R-Oc-A committed:
Adding local gravity and local temperature calculators.

R-Oc-A committed:
First complete version of the temp_name_lib

R-Oc-A committed:
Working on pulstar binary. 50% completion on writing the parsing input file routines

R-Oc-A committed:
Continue development for the parsing function of the input file. Still not finished. This is just a check point.

R-Oc-A committed:
Adding tests for parsing routines. Parsing individual values(cell values) done. Left to do: tests for strings parsing and tests for file parsing.

R-Oc-A committed:
Reestructuring Utils from the pulstar code, parsing functions (value, string, file) will be distributed into single modules

R-Oc-A committed:
Parsing string tests done; writing tests for parsing files

R-Oc-A committed:
Finished writing the first set of parsing tests. Might add some more in the future. So far, as long as the input files are well written, the program pulstar is able to initialize the setup.

R-Oc-A committed:
Continuing to write the pulstar main function.

R-Oc-A committed:
Continuing development of main for pulstar code

R-Oc-A committed:
Adding print report file; missing just the writing functions.

R-Oc-A committed:
Uploading input files and a running pulstar version of the code. The file writing methods and/or functions are to be implemented yet.

R-Oc-A committed:
Finished first version of the pulstar code. It now uses the polars crate and writes the information of the spherical grid on a parquet file

R-Oc-A committed:
First commit for the profile code; opening of the pulstar's output parquet file done; Working now on traversing the intensity grids using polars

R-Oc-A committed:
Added a  function that selects the rectangle in the temperature_loggravity space where you would do lineal interpolation. This function relies heavily on polars for the query

R-Oc-A committed:
continuing adding methods for profile

R-Oc-A committed:
Beginning to reestructure profile before advancing further. This is an inbetween commit

R-Oc-A committed:
Continuation of profiles reestructure

R-Oc-A committed:
finished adding computations of flux and continuum intensity accross selected wavelengths for a given theta-phi cell

R-Oc-A committed:
developping functions for writting the parquet output file

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+use std::fs::File;
+use std::sync::Arc;
+pub struct config{
+    pub lambda_0:f64,
```
You mean lambda_begin and lambda_end?

R-Oc-A replied:
I'll adjust the name

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+pub struct config{
+    pub lambda_0:f64,
+    pub lambda_f:f64,
+    pub delta_lbd:f64,
```
No abbreviations like `lbd` please, they make it less easy to maintain.

R-Oc-A replied:
I'll adjust the name

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+    pub grids:Vec<GridFeatures>,//<--effective temperature,log gravity;name_of_grid_file
+    pub lambda:Vec<f64>,//<-relevant wavelengths of grids
+
+    pub t_eff_useful_index: usize,
```
What does "useful" mean? It 's not clear from the variable name.

R-Oc-A replied:
I defined it as  written on your c++ code. This index was used on the profile code to put dabs on an intensity grid file so that it wouldn't be necessary to repeat a query. It's not useful anymore

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+pub struct GridFeatures{
+    pub t_eff:f64,
+    pub log_g:f64,
+    pub file_name:String,
```
Why is the filename included in the GridFeatures? If in the future you would pull the info from a DB, this would need to be re-implemented.

R-Oc-A replied:
I agree, it will be changed with a DB in mind.  
On that note I believe currently, for the profile code, I don't use this data structure...

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+    coschi:f64,
+    temperature:f64,
+    log_gravity:f64,
+    doppler_shift:f64,
```
Although clear, your function naming is unconventional in the rust community. 
Please document you function arguments! They should contain at least the units.

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+}
+
+//function that appends doppler shift
+pub fn append_doppler_shift(df:LazyFrame)// I take ownership of the data frame since I will produce a new one and want the old one to be dropped after appending
```
The Doppler shift is delta_lambda, and is expressed in nm. What you call the Doppler shift is actually the relative doppler wavelength shift and is dimensionless. 
This function should not be public. It's a helper function, that is not particularly useful for anything else that this particular dataframe.

R-Oc-A replied:
I do need to restructure  the access of certain functions. I might need to group several functions on a separate module. 
I'll change the name to something that can express "relative Doppler wavelength shift" without being too verbose.

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+		Field::new("dc".into(), DataType::Float64)
+    ]); 
+   
+    /*let parse_options = CsvParseOptions{
```
Try to refrain from leaving commented out sections of code. For your successors this will cause problems as they don't know why it is commented out, or whether you forgot to uncomment it.

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+/// Input: 
+/// Output: (intens_result,cont_result) where intens is the total intensity; cont is the continuum intensity
+/// 
+pub fn interpolate(
```
IMHO you can make a separate module (e.g. "InterpolatingFunction") that takes a grid to construct it and then behaves like a function so that you can `f(x,y,z)` in a transparant way. This will for sure be more broadly useful than only intensity grids.

R-Oc-A replied:
I agree, rn the interpolate function is contained on a module called "interpolate"

JorisDeRidder commented on profile/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+
+
+
+fn search_geq(vector:&[f64],key:f64)-> usize {
```
It's not clear at all what this function does, or why it's useful.

R-Oc-A replied:
This is a function that applies the [binary search](https://en.wikipedia.org/wiki/Binary_search) for the first great or equal member of a sorted vector.  It is an internal function and it's broadly applied to any f64 vector so I'm not sure how to document it since it's not meant to be used by the end-user.
I use it to search for the index of the closest  wavelength on the intensity grid file before interpolating.
Will change the name and add a description in the following days.

JorisDeRidder commented on Cargo.toml:
```diff
@@ -1,10 +1,3 @@
-
+[workspace]
+resolver = "3"
+members = [ "profile","pulstar", "temp_name_lib"]
```
We should consider putting pulstar and profile in two different repositories. Profile is in fact more broadly useful than just pulsating stars. It could, for example, also be used if you want to simulated double-lined binaries.

JorisDeRidder commented on profile/src/main.rs:
```diff
@@ -0,0 +1,152 @@
+use std::{env,fs::File};
+use polars::prelude::{LazyFrame,LazyGroupBy};
+use polars::prelude::*;
+use temp_name_lib::type_def::{CLIGHT, N_FLUX_POINTS};
```
I would put physical constants in a file `physical_constants`. I'm not sure why `N_FLUX_POINTS` needs to be hardcoded. That seems very ugly.

R-Oc-A replied:
I would like to discuss the constants with you as well. I thought that N_Flux_points was so that the number of flux points requested was lower than available in each Intensity Grid file.

JorisDeRidder commented on profile/src/main.rs:
```diff
@@ -0,0 +1,152 @@
+   let capacity = ((profile_config.lambda_f - profile_config.lambda_0)
+        /profile_config.delta_lbd).floor() as usize + 1usize ;
+
+    if capacity >= N_FLUX_POINTS as usize {panic!("Error, too many flux points requested.")}
```
Where too many points are requested is entirely dependent on the amount of memory you have. At the HPC we have computing nodes with a specially large RAM.

JorisDeRidder commented on profile/src/main.rs:
```diff
@@ -0,0 +1,152 @@
+    let extract_time_series = tf.column("time").unwrap();
+    let time_points:Vec<f64> = extract_time_series.f64().unwrap().into_iter().flatten().collect();
+
+    for phases in time_points.iter() {
```
You mean `for phase in `, not `for phases in`, right?

JorisDeRidder commented on profile/src/main.rs:
```diff
@@ -0,0 +1,152 @@
+
+    for phases in time_points.iter() {
+        //get vector of theta points
+        let expr = col("time").eq(lit(*phases));
```
`expr` is not a great variable name. How about `current_phase`. Then an expression like `lf.filter(current_phase)` becomes much more readable.

JorisDeRidder commented on profile/src/main.rs:
```diff
@@ -0,0 +1,152 @@
+            let pf_visible =phi_frame.filter(expr);
+            let pf = append_doppler_shift(pf_visible).collect().unwrap();
+            let doppler_shift_series = pf.column("doppler shift").unwrap();
+            let area_series = pf.column("area").unwrap();
```
In general in rust you don't add the type of the variable (e.g. `series`) to the name. Instead of `area_series` you would rather state `areas`.

R-Oc-A replied:
I see...

JorisDeRidder commented on pulstar/input_pulstar.txt:
```diff
@@ -0,0 +1,64 @@
+# This is an input file for the program PULSTAR
```
I would REALLY appreciate if you would reformat this in a yaml file.

R-Oc-A replied:
Of course! I agree 100%

JorisDeRidder commented on pulstar/src/main.rs:
```diff
@@ -0,0 +1,276 @@
+            velocity_projection::{project_vpuls, project_vrot}
+            }
+        },
+    type_def::{PHI_STEP, THETA_STEP,CYCLI2RAD,RADIUSSUN},
```
`PHI_STEP` and `THETA_STEP` should not be hardcoded.

JorisDeRidder commented on pulstar/src/main.rs:
```diff
@@ -0,0 +1,276 @@
+    let parameter_file = &env_args[1];
+    let time_points_file = &env_args[2];
+
+    println!("--------------------");
```
Banners are so 90's...  :-)

R-Oc-A replied:
well, I am a 90's kid 

On a serious note, I would love to discuss how and what to prompt status of the calculations.

JorisDeRidder commented on pulstar/src/utils.rs:
```diff
@@ -0,0 +1,76 @@
+    BOOL(bool),
+}
+
+enum InputLines{
```
All of this Un-Rusty code can be avoided if you convert the inputfile to YAML format.

JorisDeRidder commented on pulstar/src/utils/parse_file.rs:
```diff
@@ -0,0 +1,347 @@
+use core::f64;
```
This whole file can be drastically shortened when the input file is converted to YAML.
It's a bit of a maintenance nightmare if other people want to add input to the input file.
The file name `parse_file.rs` does not reveal which file it concerns.

JorisDeRidder commented on pulstar/src/utils/parse_file/parse_file_tests.rs:
```diff
@@ -0,0 +1,370 @@
+use super::*;
```
This file can be deleted if you convert the input file to YAML.

JorisDeRidder commented on pulstar/src/utils/parse_value.rs:
```diff
@@ -0,0 +1,83 @@
+use super::InputKind;
```
This file can be removed if you convert the inputfile to YAML.

JorisDeRidder commented on pulstar/src/utils/write_grid_data.rs:
```diff
@@ -0,0 +1,96 @@
+
+use crate::utils::PulstarConfig;
+
+pub struct RasterStarOutput{//[Ricardo]: This is redundant, but parquet handles value repetition. 
```
Do you mean by `Raster` a Serialized grid?

R-Oc-A replied:
yes

JorisDeRidder commented on temp_name_lib/src/joris_math/ref_frame_convrs.rs:
```diff
@@ -0,0 +1,82 @@
+use crate::type_def::{VectorComponents, VectorBase};
```
I would rename this file into `reference_frames.rs`.

R-Oc-A replied:
I see the point of that

JorisDeRidder commented on temp_name_lib/src/star_physics/intensity.rs:
```diff
@@ -0,0 +1,71 @@
+}
+
+
+pub fn hunt(vec:&[f64],x:f64)->Result<usize,MathErrors>{
```
I would be surprised if there is no existing `hunt()` functionality in Rust that you can re-use.

R-Oc-A replied:
I actually erased this function from pulstar as it is only used for profile on the c++ version. 
I think that the proposed one, with the temporal name "search_geq", is useful as it is, because it's really short, and uses the well known search algorithm

JorisDeRidder commented on temp_name_lib/src/star_physics/intensity.rs:
```diff
@@ -0,0 +1,71 @@
+
+    lambda_lower:f64,
+    lambda_upper:f64,
+    limb_darkening_coef:na::Vector4<f64>,
```
Do you really want to hardcode that you have 4 limbdarkening coefficients? If you later on want to improve, you will need to change it everywhere in the code.

R-Oc-A replied:
I did not use this. So I think this three variables may be deleted.

JorisDeRidder commented on temp_name_lib/src/star_physics/intensity.rs:
```diff
@@ -0,0 +1,71 @@
+    lambda_upper:f64,
+    limb_darkening_coef:na::Vector4<f64>,
+    grid_status:GridPointStatus,
+    kind:IntensityType,
```
Why is it useful to have the intensity type as part of the struct? Does the code use this information dynamically? I would have expected somewhere in the code
`continuum_intensities = `
and
`total_intensities = `
so that it's immediately clear what kind of grid it concerns.

R-Oc-A replied:
Intensity.rs is not useful for pulstar and I wrote a module for profile so I think this will not be used.

JorisDeRidder commented on temp_name_lib/src/star_physics/intensity.rs:
```diff
@@ -0,0 +1,71 @@
+    pub fn read_file(){}
+    pub fn read_info(){}
+    
+    pub fn set_wavelength_range(&mut self,lower:f64,upper:f64){
```
Why this function? The intensities are read from a file. So the wavelength range is clear from what's in the file.

R-Oc-A replied:
I agree.

JorisDeRidder commented on temp_name_lib/src/star_physics/intensity.rs:
```diff
@@ -0,0 +1,71 @@
+
+impl Intensity{
+
+    pub fn read_file(){}
```
The Rustacean way to name this is `from_file()` so that you can write 
`intensities = Intensity::from_file(path)`

Note, however, that what you want is what is called an `InterpolatingFunction`: an object that behaves like a function and does not need to make the user know that it has an underlying grid.

JorisDeRidder commented on temp_name_lib/src/star_physics/local_values.rs:
```diff
@@ -0,0 +1,73 @@
+//This function calculates the local time dependent values of the
+//Eigenfunctions (Local gravity or Temperature) in units of some
+//base value. Internal use only.
+pub fn local_value(value: &[Eigenfunctions],
```
`local_value()` is a poor function name.

JorisDeRidder commented on temp_name_lib/src/star_physics/local_values/gravity.rs:
```diff
@@ -0,0 +1,40 @@
+use super::local_value;
```
The setup `star_physics/local_values/*.rs` seems overly complicated. All of these functions could be in one file only.

JorisDeRidder commented on temp_name_lib/src/star_physics/velocity/pulsation.rs:
```diff
@@ -0,0 +1,49 @@
+use crate::joris_math::spherical_harmonics::d_plmcos_dtheta::deriv1_plmcos_dtheta;
```
Also here I don't understand the usefulness of the complicated `star_physics/velocity/*.rs` setup.

R-Oc-A replied:
First I was trying to order the sets of data structures, functions, methods and so on from the pulstar project on c++
After trying to read and somehow get a big picture on everything I realized it was easier to start rewriting the code and decide how to group the abstractions on the fly. I'm not so sure that having a separate module for star physics is the right approach and also the shared library "temp_name_lib" is useless in my opinion since most of the methods in there are only used for the pulstar code. 

Aside from that. I want to argue in favor of having pulstar and profile on the same workspace as that enables the conflict resolution on both codes.

JorisDeRidder commented on temp_name_lib/src/type_def.rs:
```diff
@@ -0,0 +1,72 @@
+//[Ricardo]: Configuration parameters of the simulation...I guess.
+// In the c++ version this are called parameters.
+#[derive(Debug,PartialEq)]
+pub struct Config{
```
The C++ version is not very good. It would make more sense to create a struct
```
pub struct PulsationMode {
    pub l: u16,
    pub m: i16,
    pub relative_amplitude: f64,
    K: f64,
}
```
which you then use as something similar to:
```
let pulsation_modes: Vec<PulsationMode>;
```

R-Oc-A replied:
It does make more sense... however implementing it would require changing in several functions of the pulstar code. 
On the other hand, it would for sure improve readability.

After I finish writing the profile code, which can produce a (meanflux, wavelengths, time) tabular data I'll spent some days regrouping things. It's important as well to implement testing as this hasn't been done for critical modules.

R-Oc-A committed:
Finished first version of functions to write the parquet output file. Will begin writing the methods to read YAML configuration files for profile code.

R-Oc-A committed:
Adding parsing of configuration files. First I will be exploring using TOML files instead of YAML since they're more natural to Rust. I'm currently working on a module called read_toml that will parse the config file into the program.

R-Oc-A committed:
I have a correct parser for a first version of the toml file. Before proceeding it's important to restructure the current version of the profile code.

R-Oc-A committed:
Rearranging the main function of profile. I will begin to declutter this main function and pass the functionalities onto the library

R-Oc-A committed:
Continuing reorganizing the profile program. finished rewriting the parser from the profile_input.toml file.

R-Oc-A committed:
Continuing reworking the intensity module. Access is now private. It will contain methods to produce a dataframe with the intensity grids that can be handled with polars

R-Oc-A committed:
Finished documenting the intensity module. The calculation of Doppler shifting of a vector of wavelengths is now moved into the main library.

R-Oc-A committed:
Finished commenting the module define_parameter_space, which is used for serching on the (temperature,log_gravity) parameter space.

R-Oc-A committed:
Finished writting comments for the interpolation module. The function that had a similar name called 'interpolate element' has been renamed to 'get_contribution_from_vertex' as it's more clear that we're using Trilinear interpolation. There has been a fix on the main interpolating function as the i and j indices now correctly run from 0 to 1 and also the values for t0, t1,t2 have been fixed.

R-Oc-A committed:
Finishing writting the comments in the utils module. There's now an explanation of how does the writting onto a parquet file works.

R-Oc-A committed:
erased unused modules or crates on main.rs. Will begin now to restructure and comment the main library

R-Oc-A committed:
Finished writing the comments for the main library lib.rs. the function 'append_doppler_shift' is now renamed as insert_col_relative_dlambda, the arguments 'doppler shift' are changed into 'relative shift'

R-Oc-A committed:
Experimenting with the documentation of the profile code. Major fix on the profile_input.toml file, profile program initializes parameters adequately

R-Oc-A committed:
pulstar change: Local_values function now returns appropriate results for log_g and Temperature

R-Oc-A committed:
Rewriting line by line the pulstar main function. Currently working on the parsing of the input files. This will be done in a similar fashion as the profile binary

R-Oc-A committed:
Pulstar_dev: Finished parsing functions to create the configuration file for the pulstar program. There's a working version for the 'pulstar_input.toml' file

R-Oc-A committed:
Pulstar_Dev: Changing the structure of the pulstar_input.toml, it now handles the Time type enum better
Beggining to drop the old parameter structure from the main file.
So far the program works addequately all the way until the beggining of the loop.

R-Oc-A committed:
Pulstar_dev: The implementation of the new config structure continues. The program works all the way up to the beggining of the  theta/phi loops

R-Oc-A committed:
Pulstar_dev: Beginning the process of migrating some of the geometry functions into the reference_frame.rs module. There's a new enumeration of a vector called Coordinates that has Spherical and Cartesian variants. The displacement function is also contained in this module with more accurate documentation

R-Oc-A committed:
Pulstar_dev: finished migrating the geometry functions over to the reference_frame.rs module. Beginning to move local pulsation values (velocity, temperature, gravity) over to thetheir own modules.

R-Oc-A committed:
Pulstar_dev: Fixed a major bug in the computation of local_logg. All of the functions to calculate local temperature and gravity have been moved to the module local_temperature_and_gravity.rs. There're three aditions to the reference frame module: two functions to calculate the amplitude of the lagrangian displacements and a method to obtain the r component of a Coordinates::spherical variant

R-Oc-A committed:
Pulstar_dev: finished migrating all of the (simple) pulsation methods and functions into the module local_pulsation_velocity.rs

R-Oc-A committed:
Pulstar_dev: finished migrating the (simple) calculation methods of the rotation velocity into the module local_pulsation_velocity.rs

R-Oc-A committed:
Pulstar_dev: Fixed some major issues regarding calculating the variations in log_g. Main function now works without the old data structures. Will begin working on two things \n 1) migrating the Coordinates structures traits and implementations into its own module. \n 2) moddifying the write_parquet module to use the new structure.

R-Oc-A committed:
Pulstar_dev: Finished migrating the implementations for coordinates into the module implementations_for_coordinates.rs

R-Oc-A committed:
Pulstar_dev: update of the writing functions. Now i need to implement so it deletes temporary parquet files, the code works stably

R-Oc-A committed:
Pulstar_dev: finished writing cleaning function for the temporary parquet files the name of the function is remove_temp_parquet_file()

R-Oc-A committed:
Profile_dev: Beginning of testing profile. Changes of gitignore file added to include a dummy rasterized_star parquet file. Currently there's an unidentified bottleneck.

R-Oc-A committed:
Profile_dev: uploading changes to the filtering expression of the intensity grid files. Now only excludes wavelengths that are out of range. in the pulstar_input.toml there are some annotations done by Conny. I think I've identified the bottleneck. for every surface cell the computation of the intensity is carried on again and again. I think there must be a faster way. I'll try to explore it

R-Oc-A committed:
Profile_dev: Beginning of big restructure of the profile code. So far it works but is slow.

R-Oc-A committed:
Profile_dev: Starting point from reorganizing the parsing of grid files. The new version will read the intensity grid files (with a .txt extension) only once at the beginning of the program and then will perform queries over Polars::DataFrames created from reading this. This commit is thus a check point of a stable version of the profile code. Also from this point on, the optimizations, to check where are the bottlenecks will be done using profilers, namely sammy which is a profiler tool for rust based on perf

R-Oc-A committed:
Profile_dev: creating submodules for the intensity module. new modules: parse_intensity_grids.rs extract_intensity_fluxes.rs

R-Oc-A committed:
Profile_dev: finished reorganizing intensity module. Now begins the modifying of the main library and the main function. reading grids from disk will be done only once

R-Oc-A committed:
Profile_dev: intensity and interpolate modules are updated. left to do is to implement changes into main function to see whether reading only once from disk accelerates computations.

R-Oc-A committed:
Profile_dev: Finished updating the main function. Beginning debuging phase of the reading once from disk version

R-Oc-A committed:
Profile_dev: Massive speedup with current setup. It is still left to gauge whether it produces  relevant results.

R-Oc-A committed:
Pulstar_dev: removing unnecessary structures and methods from temp_name_lib. joris_math module has changed name into math_module. Pulstar has been fixed, the rust compiler doesn't show warnings at this stage

R-Oc-A committed:
Pulstar_dev: Uploading a checkpoint of the pulstar binary. Although the program compiles and adequately implements time advancing, parsing the input file needs to be adjusted.

R-Oc-A committed:
Pulstar_dev: Finished writing the new parsing methods for the pulstar binary. Will begin to addres that there's currently no change on the pulsation phases

R-Oc-A committed:
Pulstar_dev: Fix-There're changes in the profiles of the pulsation phases. Next to do is to decide whether it's convenient to have separate files for the sphere's data or to have them as it is in one single file.

R-Oc-A committed:
Profile_dev: erasing unused code for  parsing intensity grid files.

R-Oc-A committed:
Profile_dev: Uploading new version to produce the output into parquet. It collect's every phase of pulsation into a single file

R-Oc-A committed:
Profile_dev: Current version of profile is stable with the default profile_input.toml file. There are left some changes to be made so that the initialization of the intensity data frames doesn't add too much overhead as it currently cannot handle 1000 wavelengths.

R-Oc-A committed:
Profile_dev: corrected a misspelling on lib.rs. On extract_intensity_fluxes.rs added a flag so that if the requested wavelength's spacing is lowresolution, then there is no second filtering prior to calculating the intensity fluxes.

R-Oc-A committed:
Profile_dev: uploading newer versions of the modules extract_inteinsity_fluxes.rs and parse_intensity_grids.rs so that it filters wavelengths if asking for low resolution

R-Oc-A committed:
Profile_dev: delleting an unused module from parse_intensity_grids.rs

R-Oc-A committed:
finished first version of the new structure of profile. It has new structures such as surface cells and fluxOfSpectra that allow to stop allocating inside loops.

R-Oc-A committed:
Rebasing of the branch that includes the multilinear interpolation done on the shared library. From this point onwards the profile binary will use that method to interpolate on the intensity grids.

R-Oc-A committed:
Uploading a commented version of the current profile binary that performs faster than using data frames. It will still be necessary to later drop the nested vectors version of the limb darkening coefficients to improve speed

R-Oc-A committed:
Deleted some comments on the main function of the profile project.

R-Oc-A committed:
Updating the .gitignore file to ignore the cheetsheet for the Samply profiler tool

R-Oc-A committed:
adding new data structures to reorganize the code. New additions:
Surface cell: contains all of the local quantities as well as the coordinates
RasterizedStar: contains a vector of surfacecells

Main program will project only this new structures. Complete version still to be finished

R-Oc-A committed:
Pusltar_dev: inbetween commit, creating the methods for updating local values over a surface cell of a star.

R-Oc-A committed:
Pulstar_dev: finished coding the methods for updating the whole rasterized star; it iterates over the surface cells and on each cell updates_local_quantities

R-Oc-A committed:
pulstar_dev: Uploading inbetween commit. There's now a less verbose structure on the main function. Although the program compiles, there seem to be only negative values for coschi

R-Oc-A committed:
Pulstar_dev: Fixed a bug in the method upload_local_quantities for a surface cell. The code properly stores the phases of the rasterized star.

R-Oc-A committed:
Pulstar_dev: Uploading a stable newer version of the pulstar code. It now has a simpler and less verbose main function. There are some methods that still need comments and documentation

R-Oc-A committed:
Pulstar_dev: Commenting this stable version of pulstar. write_grid_data.rs has readjusted the visibility of the RasterizedStarOutput struct to private. .gitignore file includes now all of *.swp files

R-Oc-A committed:
Multilinear-interpolation: inbetween commit. There's still lot's of methods and structures to program

R-Oc-A committed:
Multilinear-interpolation: inbetween commit. Midway point

R-Oc-A committed:
Multilinear-interpolation: Finished coding multilinear interpolation. Steps to left to do: Set up a unit test for the multilinear interpolation method; create a mehtod for filling the hypercube on the parameter space

R-Oc-A committed:
Multilinear-interpolation: Inbetween commit. 1D version of linear interpolation done. Next is to create tests that push this module to work with Nadya's Spectral grids.

R-Oc-A committed:
Multilinear-interpolation: inbetween commit. I Finally decided on using ndarray to handle the memory layout of the specific intensity grids. Right now I'm coding a test with a sample of Nadya's grids. It's important to notice that she has the mu values in descending order.

R-Oc-A committed:
Multilinear-interpolation: Finished coding the first part of the test which was creating a sample of the data contained in the intensity grids. I now have to manually code the multilinear interpolation to compare the results

R-Oc-A committed:
Multilinear-interpolation: Finished conding the test with Nadya's sample grid. It passed. Documenting left
to do

R-Oc-A committed:
Multi-linear interpolation: There was a mistake on the multilinear test. Added now a check to see if the interpolating point is located inside the hypercube. Finished documenting the methods

R-Oc-A committed:
Multilinear interpolation: Added a new test to see if the multilinear interpolation method works on 1D that it was successful. Pending the 2D case

R-Oc-A committed:
Profile_dev_with_nadyalike_grid: First commit. Uploading updated cargo.toml from profile binary in order to include ndarray.

R-Oc-A committed:
profile_dev_with_nadyalike_grid: adding a new data structure that will hold the information of the specific intensity grids as a ndarray::Array3<f64>

R-Oc-A committed:
profile_dev_with_nadyalike_grid: inbetween commit

R-Oc-A committed:
profile_dev_with_nadyalike_grid: finished adding new implementations in order to parse csv data into ndarray::Array3; parsing either Joris or Nadyas grids is left to be done

R-Oc-A committed:
profile_dev_with_nadyalike_grid: inbetween commit

R-Oc-A committed:
profile_dev_with_nadyalike_grid: inbetween commit. SpectralGrid will be a data structure rather than an enumeration. Joris like grids will be transformed to arrays that have specific intensities for 7 mu values very much like Nadya's grids and moving forward the way to go will be to ask all of the user defined grids to be transformed into this setup.

R-Oc-A committed:
profile_nadya: yeah lot of work to do:

R-Oc-A committed:
profile_dev_with_nadyalike_grid: finished adding all the implementations. All of the grids will hold nadyalike grids structure. Discriminating cells with mu values below 0.26 and grids are left to be done

R-Oc-A committed:
profile_dev_with_nadyalike_grid: created an adequate profile_input.toml file, so far the spectral grid was done correctly I presume. Beggining debugging the extract_flux_from_cell

R-Oc-A committed:
profile with nadyalike grid: does not crash with joris grids but results are not similar at all. Must include how to handle for high mu values. Also it might be interesting to check if it's possible to translate from nadya to joris or thing something along those lines

R-Oc-A committed:
profile_dev_with_nadyalike_grid: stable version so far. results are comparable with Joris's ones

R-Oc-A committed:
Fixing the nomenclature inside of the saving function. Mean flux changed to the actual term of normalized flux

R-Oc-A committed:
deconstructing the main() function of the profile crate in order to move it into a separate function that will be called by the python interface

R-Oc-A committed:
Testing output to check for differences with the un-refactored version

R-Oc-A committed:
Deconstructing profile main function done

R-Oc-A committed:
merging of main and pulstar_dev

R-Oc-A committed:
Merge branch 'main' into refactoring_single_lib
" Previous commit of this branch had an outdated version of pulstar"

R-Oc-A committed:
developping python interface

R-Oc-A committed:
Python_Module_dev:running pulstar from the pulstar_py...done

R-Oc-A committed:
Python_module_dev: small modifications.

R-Oc-A committed:
Working version that is callable from python. From here on I'll be working on creating and parsing the configuration files from python to Rust

R-Oc-A committed:
pulstar and profile receive toml files as input. We'll be creating a python script to produce this toml strings instead of generating objects from python as the only relevant things for the project is to retreive specific intensities.

R-Oc-A committed:
Adding fully programmed python script to create both pulstar and profile toml strings. Moddified version of profile_main that now takes a toml string as input

R-Oc-A committed:
modified pulstar and profile main functions. They now work together and produce a parquet file at the end with the intensity written on a parquet file

R-Oc-A committed:
Able to obtain and open a DataFrame from rust in python. I had to downgrade pyo3 so that it was compatible with polars version 0.49.1 It is left to do to update pulstar and profile to have the newest polars and pyo3 crate versions. I'll be doing this in later stages of the project. In the mean time I'll proceed to attend both the implementation of TAR and the fitting strategies.

R-Oc-A committed:
Stable point for pulstar, profile, and python interface. Will begin coding a TAR implementation that relies on Chebyshev polynomials. I will later develop an expansion of Hough in terms of Spherical Harmonics

R-Oc-A committed:
Hough's functions using the collocation method developed by Vincent Prat and implemented in Rust by me.

R-Oc-A committed:
pulstar_dev: Adding structure for implementing the different rotation regimes.

R-Oc-A committed:
pulstar_dev: uploaded new version of pulstar_input.toml.

R-Oc-A committed:
pulstar_dev: moddifying python interface. Now it's possible to call pulstar and profile separately

R-Oc-A committed:
pulstar_dev: Documented implementation of Hough function

R-Oc-A committed:
pulstar_dev: inbetween commit output of houghs function is now a regular vector because parsing ndarray seems to be problematic. I'm halfway done with implementing pulsational displacements using tar. I need to have a consistent prograde/retrograde for everything. TAR uses q*m>0 retrograde and I think the non rotating and the perturbative coriolis approach are the other way around.

R-Oc-A committed:
pulstar_dev: uploading rotation displacements using TAR

R-Oc-A committed:
pulstar_dev: uploading recent changes of implementation of tar in pulstar part of the project. Non rotating case works like it always has; the TAR implementation is not finished yet. I've noticed that the expected eigenvalue differs from the python implementation.

R-Oc-A committed:
TAR: corrected values on the implementation of houghs functions

R-Oc-A committed:
TAR: some other corrected values on the implementation of houghs functions

R-Oc-A committed:
TAR: standarization of Hough functions, they are no longer dependent on the number of collocation points. Previously the maximum amplitude decreased with the number of collocation points in order to maintain that the L2-norm remains equal to 1.

R-Oc-A committed:
pulstar_dev: tar implementation now has at list 400 collocation points for the houghs functions. A new way of getting the index of the function's arrays is given

R-Oc-A committed:
TAR: revised version of hough's functions.
